src/annotate.cpp

```diff
@@ -0,0 +1,524 @@
+        return overlays[ws.x][ws.y];
+    }
+
+    wf::signal_connection_t viewport_changed{[this] (wf::signal_data_t *data)
+    {
+        output->render->damage_whole();
+    }};
+
+    wf::button_callback draw_begin = [=] (uint32_t, int x, int y)
+    {
+        grab_point = last_cursor = wf::get_core().get_cursor_position();
+
+        grab();
+
+        return true;
+    };
+
+    void draw_end()
+    {
+        auto& ol = get_current_overlay();
+
+        overlay_destroy(shape_overlay);
+        ungrab();
+
+        if (std::string(method) == "line")
+        {
+            cairo_draw_line(ol, wf::get_core().get_cursor_position());
+        }
+        else if (std::string(method) == "rectangle")
+        {
+            cairo_draw_rectangle(ol, last_cursor);
+        }
+        else if (std::string(method) == "circle")
+        {
+            cairo_draw_circle(ol, last_cursor);
+	}
+    }
+
+    void pointer_moved(wlr_event_pointer_motion* ev)
+    {
+        auto current_cursor = wf::get_core().get_cursor_position();
+        auto& ol = get_current_overlay();
+
+        if (std::string(method) == "draw")
+        {
+            cairo_draw(ol, last_cursor, current_cursor);
+        }
+        else if (std::string(method) == "line")
+        {
+            cairo_draw_line(shape_overlay, current_cursor);
+        }
+        else if (std::string(method) == "rectangle")
+        {
+            cairo_draw_rectangle(shape_overlay, current_cursor);
+        }
+        else if (std::string(method) == "circle")
+        {
+            cairo_draw_circle(shape_overlay, current_cursor);
+        }
+
+        last_cursor = current_cursor;
+
+        connect_ws_stream_post();
+    }
+
+    void deactivate_check()
+    {
+        bool all_workspaces_clear = true;
+        auto wsize = output->workspace->get_workspace_grid_size();
+
+        for (int x = 0; x < wsize.width; x++)
+        {
+            for (int y = 0; y < wsize.height; y++)
+            {
+                auto& ol = overlays[x][y];
+                if (ol.cr)
+                {
+                    all_workspaces_clear = false;
+                    x = wsize.width;
+                    break;
+                }
+            }
+        }
+
+        if (all_workspaces_clear)
+            disconnect_ws_stream_post();
+    }
+
+    void overlay_clear(anno_ws_overlay& ol)
+    {
+        if (!ol.cr)
+            return;
+
+        cairo_clear(ol.cr);
+    }
+
+    void overlay_destroy(anno_ws_overlay& ol)
+    {
+        if (!ol.cr)
+            return;
+
+        ol.texture->release();
+        ol.texture.reset();
+        cairo_surface_destroy(ol.cairo_surface);
+        cairo_destroy(ol.cr);
+        ol.cr = nullptr;
+    }
+
+    void clear()
+    {
+        auto& ol = get_current_overlay();
+
+        overlay_destroy(ol);
+        deactivate_check();
+
+        output->render->damage_whole();
+    }
+
+    wf::signal_connection_t workarea_changed{[this] (wf::signal_data_t *data)
```

Comment:
This is probably the wrong signal for this, but the idea is that the texture is the size of the output, and when the output size changes, it makes sense to destroy the texture so it will be recreated with the new size.